File: packages/geometry-utils/src/clipper-wrapper.ts
import type { BoundRect, NestConfig, Polygon, PolygonNode } from './types';
import { getPolygonNode } from './helpers';
import { PointF32, PolygonF32 } from './geometry';
import { clean_node_inner_wasm, offset_node_inner_wasm } from 'wasm-nesting';
export default class ClipperWrapper {

    private polygon: Polygon<Float32Array>;

    constructor() {
        this.polygon = new PolygonF32();
    }

    public generateBounds(memSeg: Float32Array, spacing: number, curveTolerance: number): {
        binNode: PolygonNode;
        bounds: BoundRect<Float32Array>;
        resultBounds: BoundRect<Float32Array>;
        area: number;
    } {
        this.polygon.bind(memSeg);

        if (this.polygon.isBroken) {
            return null;
        }

        const binNode: PolygonNode = getPolygonNode(-1, memSeg);
        const bounds: BoundRect<Float32Array> = this.polygon.exportBounds();

        this.cleanNode(binNode, curveTolerance);
        this.offsetNode(binNode, -1, spacing, curveTolerance);

        this.polygon.bind(binNode.memSeg);
        this.polygon.resetPosition();

        const resultBounds = this.polygon.exportBounds();
        const area: number = this.polygon.area;

        return { binNode, bounds, resultBounds, area };
    }

    public generateTree(memSegs: Float32Array[], spacing: number, curveTolerance: number): PolygonNode[] {
        const point: PointF32 = PointF32.create();
        const trashold = curveTolerance * curveTolerance;
        const nodes: PolygonNode[] = [];
        const nodeCount: number = memSegs.length;
        let memSeg: Float32Array = null;
        let node: PolygonNode = null;
        let i: number = 0;

        for (let i = 0; i < nodeCount; ++i) {
            memSeg = memSegs[i];
            node = getPolygonNode(i, memSeg);

            this.cleanNode(node, curveTolerance);

            this.polygon.bind(node.memSeg);

            if (this.polygon.isBroken || this.polygon.absArea <= trashold) {
                console.warn('Can not parse polygon', i);
                continue;
            }

            nodes.push(node);
        }

        // turn the list into a tree
        this.nestPolygons(point, nodes);

        this.offsetNodes(nodes, 1, spacing, curveTolerance);

        this.simplifyNodes(nodes);

        return nodes;
    }


    private simplifyNodes(nodes: PolygonNode[]): void {
        const nodeCount: number = nodes.length;
        let size: number = 0;
        let i: number = 0;
        let j: number = 0;

        for (i = 0; i < nodeCount; ++i) {
            size = nodes[i].memSeg.length;

            this.simplifyNodes(nodes[i].children);

            for (j = 0; j < size; ++j) {
                nodes[i].memSeg[j] = Math.round(nodes[i].memSeg[j] * 100) / 100;
            }
        }
    }

    // Main function to nest polygons
    private nestPolygons(point: PointF32, nodes: PolygonNode[]): void {
        const parents: PolygonNode[] = [];
        // assign a unique id to each leaf
        let nodeCount: number = nodes.length;
        let outerNode: PolygonNode = null;
        let innerNode: PolygonNode = null;
        let isChild: boolean = false;

        for (let i = 0; i < nodeCount; ++i) {
            outerNode = nodes[i];
            isChild = false;
            point.fromMemSeg(outerNode.memSeg);

            for (let j = 0; j < nodeCount; ++j) {
                innerNode = nodes[j];
                this.polygon.bind(innerNode.memSeg);

                if (j !== i && this.polygon.pointIn(point)) {
                    innerNode.children.push(outerNode);
                    isChild = true;
                    break;
                }
            }

            if (!isChild) {
                parents.push(outerNode);
            }
        }

        for (let i = 0; i < nodeCount; ++i) {
            if (parents.indexOf(nodes[i]) < 0) {
                nodes.splice(i, 1);
                --nodeCount;
                --i;
            }
        }

        const parentCount: number = parents.length;
        let parent: PolygonNode = null;

        for (let i = 0; i < parentCount; ++i) {
            parent = parents[i];

            if (parent.children) {
                this.nestPolygons(point, parent.children);
            }
        }
    }

    private offsetNodes(nodes: PolygonNode[], sign: number, spacing: number, curveTolerance: number): void {
        const nodeCont: number = nodes.length;
        let node: PolygonNode = null;
        let i: number = 0;

        for (i = 0; i < nodeCont; ++i) {
            node = nodes[i];
            this.offsetNode(node, sign, spacing, curveTolerance);
            this.offsetNodes(node.children, -sign, spacing, curveTolerance);
        }
    }

    private offsetNode(node: PolygonNode, sign: number, spacing: number, curveTolerance: number): void {
        node.memSeg = offset_node_inner_wasm(node.memSeg, sign, spacing, curveTolerance);
    }


    private cleanNode(node: PolygonNode, curveTolerance: number): void {
        const res = clean_node_inner_wasm(node.memSeg, curveTolerance);

        if (res.length) {
            node.memSeg = res;
        }
    }

    public static CLIPPER_SCALE: number = 100;

    public static AREA_TRASHOLD: number = 0.1 * ClipperWrapper.CLIPPER_SCALE * ClipperWrapper.CLIPPER_SCALE;

    public static CLEAN_TRASHOLD: number = 0.0001 * ClipperWrapper.CLIPPER_SCALE;
}

File: packages/wasm-nesting/src/clipper_wrapper.rs
use crate::geometry::point::Point;
use crate::geometry::polygon::Polygon;
use crate::nesting::polygon_node::PolygonNode;
use crate::utils::number::Number;

/// Clean a polygon node by removing degenerate points
fn clean_node(node: &mut PolygonNode, curve_tolerance: f64) {
    let cleaned = crate::clean_node_inner(&node.mem_seg, curve_tolerance);
    if !cleaned.is_empty() {
        node.mem_seg = cleaned.into_boxed_slice();
        node.seg_size = node.mem_seg.len();
    }
}

/// Apply offset to a polygon node
fn offset_node(node: &mut PolygonNode, sign: i32, spacing: i32, curve_tolerance: f64) {
    let offset_result = crate::offset_node_inner(&node.mem_seg, sign, spacing, curve_tolerance);
    if !offset_result.is_empty() {
        node.mem_seg = offset_result.into_boxed_slice();
        node.seg_size = node.mem_seg.len();
    }
}

/// Recursively simplify nodes by rounding coordinates to 2 decimal places
fn simplify_nodes(nodes: &mut [PolygonNode]) {
    for node in nodes.iter_mut() {
        // Round all coordinates to 2 decimal places
        for value in node.mem_seg.iter_mut() {
            *value = (*value * 100.0).round() / 100.0;
        }

        // Recursively simplify children
        simplify_nodes(&mut node.children);
    }
}

/// Recursively nest polygons into a parent-child tree structure
fn nest_polygons(nodes: &mut Vec<PolygonNode>) {
    let node_count = nodes.len();
    let mut parents = Vec::new();
    let mut test_point = Point::<f32>::new(None, None);
    let mut polygon = Polygon::<f32>::new();

    // Find parent-child relationships
    for i in 0..node_count {
        let mut is_child = false;

        // Get the first point of the outer polygon
        unsafe {
            test_point.set(nodes[i].mem_seg[0], nodes[i].mem_seg[1]);
        }

        // Check if this polygon is inside any other polygon
        for j in 0..node_count {
            if i == j {
                continue;
            }

            let point_count = nodes[j].mem_seg.len() >> 1;
            unsafe {
                polygon.bind(nodes[j].mem_seg.clone(), 0, point_count);

                if let Some(true) = polygon.point_in(&test_point as *const Point<f32>, None) {
                    is_child = true;
                    break;
                }
            }
        }

        if !is_child {
            parents.push(i);
        }
    }

    // Separate children from parents
    let mut children_map: Vec<(usize, Vec<usize>)> = Vec::new();

    for i in 0..node_count {
        if parents.contains(&i) {
            continue;
        }

        // Get the first point of the child polygon
        unsafe {
            test_point.set(nodes[i].mem_seg[0], nodes[i].mem_seg[1]);
        }

        // Find which parent contains this child
        for &parent_idx in &parents {
            let point_count = nodes[parent_idx].mem_seg.len() >> 1;
            unsafe {
                polygon.bind(nodes[parent_idx].mem_seg.clone(), 0, point_count);

                if let Some(true) = polygon.point_in(&test_point as *const Point<f32>, None) {
                    // Find or create entry for this parent
                    if let Some(entry) = children_map.iter_mut().find(|(p, _)| *p == parent_idx) {
                        entry.1.push(i);
                    } else {
                        children_map.push((parent_idx, vec![i]));
                    }
                    break;
                }
            }
        }
    }

    // Build the tree structure
    let mut result = Vec::new();
    let mut used_indices = Vec::new();

    for &parent_idx in &parents {
        let mut parent_node = nodes[parent_idx].clone();

        // Find children for this parent
        if let Some((_, child_indices)) = children_map.iter().find(|(p, _)| *p == parent_idx) {
            let mut children = Vec::new();
            for &child_idx in child_indices {
                children.push(nodes[child_idx].clone());
                used_indices.push(child_idx);
            }

            // Recursively nest children
            nest_polygons(&mut children);
            parent_node.children = children;
        }

        result.push(parent_node);
        used_indices.push(parent_idx);
    }

    *nodes = result;
}

/// Recursively apply offset to nodes and their children with alternating signs
fn offset_nodes(nodes: &mut [PolygonNode], sign: i32, spacing: i32, curve_tolerance: f64) {
    for node in nodes.iter_mut() {
        offset_node(node, sign, spacing, curve_tolerance);
        offset_nodes(&mut node.children, -sign, spacing, curve_tolerance);
    }
}

/// Main function to generate polygon tree from flat arrays
pub fn generate_tree(
    values: &[f32],
    sizes: &[u16],
    spacing: i32,
    curve_tolerance: f64,
) -> Vec<PolygonNode> {
    let threshold = curve_tolerance * curve_tolerance;
    let mut nodes = Vec::new();
    let mut offset = 0;

    // Parse polygons from flat arrays
    for (i, &size) in sizes.iter().enumerate() {
        let seg_size = (size as usize) << 1;
        if offset + seg_size > values.len() {
            break;
        }

        let mem_seg = values[offset..offset + seg_size].to_vec();
        let mut node = PolygonNode::new(i as i32, 0.0, mem_seg);

        // Clean the polygon
        clean_node(&mut node, curve_tolerance);

        // Calculate absolute area
        let abs_area = f32::abs_polygon_area(&node.mem_seg);

        if abs_area <= threshold {
            eprintln!("Cannot parse polygon {}", i);
            offset += seg_size;
            continue;
        }

        nodes.push(node);
        offset += seg_size;
    }

    // Build the tree structure
    nest_polygons(&mut nodes);

    // Apply offset with alternating signs
    offset_nodes(&mut nodes, 1, spacing, curve_tolerance);

    // Simplify coordinates
    simplify_nodes(&mut nodes);

    nodes
}
